File: occasum-core/src/main/java/pw/stamina/occasum/properties/collection/ListProperty.java
package pw.stamina.occasum.properties.collection;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class ListProperty<T> extends AbstractCollectionProperty<T, List<T>> {

    ListProperty(String name, List<T> elements, CollectionPropertyAdapter<T> adapter) {
        super(name, elements, adapter);
    }

    @Override
    public List<T> getElements() {
        return Collections.unmodifiableList(elements);
    }

    public static <T> ListProperty<T> array(String name, CollectionPropertyAdapter<T> adapter) {
        return new ListProperty<>(name, new ArrayList<>(), adapter);
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/properties/collection/ListProperty.kt
package pw.stamina.occasum.properties.collection

import java.util.*

class ListProperty<T> internal constructor(
        name: String,
        elements: List<T>,
        adapter: CollectionPropertyAdapter<T>)
    : AbstractCollectionProperty<T, List<T>>(name, elements, adapter) {

    override val elements = elements
        get() = Collections.unmodifiableList(field)

    companion object {

        fun <T> array(name: String, adapter: CollectionPropertyAdapter<T>): ListProperty<T> {
            return ListProperty(name, ArrayList(), adapter)
        }
    }
}
